// ngxs-reset-plugin/esm5/lib/reset.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { APP_INITIALIZER, NgModule } from '@angular/core';
import { NGXS_PLUGINS } from '@ngxs/store';
import { noop } from './internals';
import { ResetHandler } from './reset.handler';
import { NgxsResetPlugin } from './reset.plugin';
import { ResetService } from './reset.service';
var NgxsResetPluginModule = /** @class */ (function () {
    function NgxsResetPluginModule() {
    }
    /**
     * @return {?}
     */
    NgxsResetPluginModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: NgxsResetPluginModule,
            providers: [
                ResetService,
                ResetHandler,
                {
                    provide: APP_INITIALIZER,
                    useFactory: noop,
                    deps: [ResetHandler],
                    multi: true,
                },
                {
                    provide: NGXS_PLUGINS,
                    useClass: NgxsResetPlugin,
                    multi: true,
                },
            ],
        };
    };
    NgxsResetPluginModule.decorators = [
        { type: NgModule }
    ];
    return NgxsResetPluginModule;
}());
export { NgxsResetPluginModule };
