// ngxs-reset-plugin/esm5/lib/reset.plugin.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { getActionTypeFromInstance, getValue, setValue } from '@ngxs/store';
import { ResetService } from './reset.service';
import { getMetaData, StateClear, StateOverwrite, StateReset, StateResetAll, } from './symbols';
var NgxsResetPlugin = /** @class */ (function () {
    function NgxsResetPlugin(resetService) {
        this.resetService = resetService;
    }
    /**
     * @private
     * @param {?} state
     * @param {?} statesToKeep
     * @return {?}
     */
    NgxsResetPlugin.prototype.clearStates = /**
     * @private
     * @param {?} state
     * @param {?} statesToKeep
     * @return {?}
     */
    function (state, statesToKeep) {
        return statesToKeep
            .map((/**
         * @param {?} meta
         * @return {?}
         */
        function (meta) { return getPath(meta); }))
            .map((/**
         * @param {?} path
         * @return {?}
         */
        function (path) { return ({
            parts: path.split('.'),
            value: getValue(state, path),
        }); }))
            .reduce((/**
         * @param {?} obj
         * @param {?} __1
         * @return {?}
         */
        function (obj, _a) {
            var parts = _a.parts, value = _a.value;
            return parts.reduceRight((/**
             * @param {?} acc
             * @param {?} part
             * @return {?}
             */
            function (acc, part) {
                var _a, _b;
                return part in obj
                    ? (_a = {},
                        _a[part] = tslib_1.__assign({}, obj[part], acc),
                        _a) : (_b = {}, _b[part] = acc, _b);
            }), value);
        }), (/** @type {?} */ ({})));
    };
    /**
     * @private
     * @param {?} state
     * @param {?} statesToOverwrite
     * @param {?} values
     * @return {?}
     */
    NgxsResetPlugin.prototype.overwriteStates = /**
     * @private
     * @param {?} state
     * @param {?} statesToOverwrite
     * @param {?} values
     * @return {?}
     */
    function (state, statesToOverwrite, values) {
        statesToOverwrite.forEach((/**
         * @param {?} meta
         * @param {?} index
         * @return {?}
         */
        function (meta, index) { return (state = setValue(state, getPath(meta), values[index])); }));
        return state;
    };
    /**
     * @private
     * @param {?} state
     * @param {?} statesToReset
     * @return {?}
     */
    NgxsResetPlugin.prototype.resetStates = /**
     * @private
     * @param {?} state
     * @param {?} statesToReset
     * @return {?}
     */
    function (state, statesToReset) {
        var _this = this;
        statesToReset.forEach((/**
         * @param {?} meta
         * @return {?}
         */
        function (meta) {
            state = setValue(state, getPath(meta), typeof meta.defaults === 'undefined' ? {} : meta.defaults);
            if (meta.children) {
                state = _this.resetStates(state, (/** @type {?} */ (meta.children.map(getMetaData))));
            }
        }));
        return state;
    };
    /**
     * @private
     * @param {?} state
     * @param {?} statesToKeep
     * @return {?}
     */
    NgxsResetPlugin.prototype.resetStatesAll = /**
     * @private
     * @param {?} state
     * @param {?} statesToKeep
     * @return {?}
     */
    function (state, statesToKeep) {
        /** @type {?} */
        var values = statesToKeep.map((/**
         * @param {?} meta
         * @return {?}
         */
        function (meta) { return getValue(state, getPath(meta)); }));
        return this.overwriteStates(this.resetService.initialState, statesToKeep, values);
    };
    /**
     * @param {?} state
     * @param {?} action
     * @param {?} next
     * @return {?}
     */
    NgxsResetPlugin.prototype.handle = /**
     * @param {?} state
     * @param {?} action
     * @param {?} next
     * @return {?}
     */
    function (state, action, next) {
        /** @type {?} */
        var type = getActionTypeFromInstance(action);
        switch (type) {
            case StateClear.type:
                state = this.clearStates(state, ((/** @type {?} */ (action))).statesToKeep);
                break;
            case StateReset.type:
                state = this.resetStates(state, ((/** @type {?} */ (action))).statesToReset);
                break;
            case StateResetAll.type:
                state = this.resetStatesAll(state, ((/** @type {?} */ (action))).statesToKeep);
                break;
            case StateOverwrite.type:
                var _a = (/** @type {?} */ (action)), statesToOverwrite = _a.statesToOverwrite, values = _a.values;
                state = this.overwriteStates(state, statesToOverwrite, values);
                break;
            default:
                break;
        }
        return next(state, action);
    };
    NgxsResetPlugin.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    NgxsResetPlugin.ctorParameters = function () { return [
        { type: ResetService }
    ]; };
    return NgxsResetPlugin;
}());
export { NgxsResetPlugin };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NgxsResetPlugin.prototype.resetService;
}
/**
 * @param {?} meta
 * @return {?}
 */
function getPath(meta) {
    return meta.path;
}
