// ngxs-reset-plugin/esm5/lib/symbols.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { isDevMode } from '@angular/core';
import { getStoreMetadata } from '@ngxs/store';
/**
 * Action to clear all state except given state(s)
 */
var StateClear = /** @class */ (function () {
    function StateClear() {
        var statesToKeep = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            statesToKeep[_i] = arguments[_i];
        }
        /** @type {?} */
        var reducer = createMetaDataListReducer(isDevMode());
        this.statesToKeep = statesToKeep.reduce(reducer, []);
    }
    StateClear.type = '@@CLEAR_STATE';
    return StateClear;
}());
export { StateClear };
if (false) {
    /** @type {?} */
    StateClear.type;
    /** @type {?} */
    StateClear.prototype.statesToKeep;
}
/**
 * Action to reset given state(s) to defaults
 */
var StateReset = /** @class */ (function () {
    function StateReset() {
        var statesToReset = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            statesToReset[_i] = arguments[_i];
        }
        /** @type {?} */
        var reducer = createMetaDataListReducer(isDevMode());
        this.statesToReset = statesToReset.reduce(reducer, []);
    }
    StateReset.type = '@@RESET_STATE';
    return StateReset;
}());
export { StateReset };
if (false) {
    /** @type {?} */
    StateReset.type;
    /** @type {?} */
    StateReset.prototype.statesToReset;
}
/**
 * Action to reset all states expect given state(s) to defaults
 */
var StateResetAll = /** @class */ (function () {
    function StateResetAll() {
        var statesToKeep = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            statesToKeep[_i] = arguments[_i];
        }
        /** @type {?} */
        var reducer = createMetaDataListReducer(isDevMode());
        this.statesToKeep = statesToKeep.reduce(reducer, []);
    }
    StateResetAll.type = '@@RESET_STATE_ALL';
    return StateResetAll;
}());
export { StateResetAll };
if (false) {
    /** @type {?} */
    StateResetAll.type;
    /** @type {?} */
    StateResetAll.prototype.statesToKeep;
}
/**
 * Action to overwrite state(s) with given value(s)
 */
var StateOverwrite = /** @class */ (function () {
    function StateOverwrite() {
        var overwriteConfigs = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            overwriteConfigs[_i] = arguments[_i];
        }
        /** @type {?} */
        var reducer = createMetaTupleReducer(isDevMode());
        var _a = tslib_1.__read(overwriteConfigs.reduce(reducer, [
            [],
            [],
        ]), 2), states = _a[0], values = _a[1];
        this.statesToOverwrite = states;
        this.values = values;
    }
    StateOverwrite.type = '@@OVERWRITE_STATE';
    return StateOverwrite;
}());
export { StateOverwrite };
if (false) {
    /** @type {?} */
    StateOverwrite.type;
    /** @type {?} */
    StateOverwrite.prototype.statesToOverwrite;
    /** @type {?} */
    StateOverwrite.prototype.values;
}
/**
 * @param {?} state
 * @param {?} devMode
 * @return {?}
 */
export function getMetaData(state, devMode) {
    /** @type {?} */
    var meta = (/** @type {?} */ (new Object(getStoreMetadata((/** @type {?} */ (state))))));
    /** @type {?} */
    var isNgxsMeta = meta.name && 'defaults' in meta;
    // Reusability Hack: devMode is number on purpose
    if (!isNgxsMeta && devMode === -2) {
        console.warn("Reset Plugin Warning: " + meta.name + " is not a state class.");
        return null;
    }
    return meta;
}
/**
 * @param {?} devMode
 * @return {?}
 */
function createMetaDataListReducer(devMode) {
    return (/**
     * @param {?} acc
     * @param {?} state
     * @return {?}
     */
    function (acc, state) {
        // tslint:disable-next-line:no-bitwise
        /** @type {?} */
        var meta = getMetaData(state, ~devMode);
        return meta ? acc.concat(meta) : acc;
    });
}
/**
 * @param {?} devMode
 * @return {?}
 */
function createMetaTupleReducer(devMode) {
    return (/**
     * @param {?} acc
     * @param {?} __1
     * @return {?}
     */
    function (acc, _a) {
        var _b = tslib_1.__read(_a, 2), state = _b[0], value = _b[1];
        // tslint:disable-next-line:no-bitwise
        /** @type {?} */
        var meta = getMetaData(state, ~devMode);
        return meta ? [acc[0].concat(meta), acc[1].concat(value)] : acc;
    });
}
